### PR TITLE
Pubsub2

### PR DESCRIPTION
**Pull request by BarneyStratford**
Introduce the pubsub module to the player_cache. This allows us to subscribe and unsubscribe to a player in constant time.

### CODE REVIEW

File: src/lib/player_cache.ts [hunk 1]
```diff
@@ -182,7 +206,7 @@ export function fetch(player_id: number, required_fields?: Array<string>): Promi
                         console.log("Batch requesting player info for", queue.map(e => e.player_id).join(','));
                     }
 
-                    get("/termination-api/players", {"ids": queue.map(e => e.player_id).join('.')})
+                    get("/termination-api/players/%%", queue.map(e => e.player_id).join(','))
```
Comment (anoek) [inline src/lib/player_cache.ts]: Picked the wrong merge here, should be
```
get("/termination-api/players", {"ids": queue.map(e => e.player_id).join('.')})
```

File: src/lib/player_cache.ts [hunk 2]
```diff
@@ -61,13 +104,7 @@ export function update(player: any, dont_overwrite?: boolean): PlayerCacheEntry
         return;
     }
 
-    let id:number = "user_id" in player ? player.user_id : player.id;
-
-    if (!id) {
```
Comment (anoek) [inline src/lib/player_cache.ts]: Lets keep this check, it helped me not too long ago

File: src/components/PlayerIcon/PlayerIcon.tsx [hunk 3]
```diff
@@ -27,48 +27,38 @@ interface PlayerIconProps {
...
-export function getPlayerIconURL(id, size): Promise<string> {{{
-    return new Promise((resolve, reject) => {
-        player_cache.fetch(id, ["icon"]).then((user) => {
-            resolve(icon_size_url(user.icon, size));
-        })
-        .catch(reject);
-    });
-}}}
+export function getPlayerIconURL(id: number, size: number): Promise<string> {
+    return player_cache.fetch(id, ["icon"]).then(user => icon_size_url(user.icon, size));
```
Comment (anoek) [inline src/components/PlayerIcon/PlayerIcon.tsx]: This does not do the same thing as the function you replaced: https://github.com/online-go/online-go.com/pull/400/files#diff-4a58653571e8b38e193938055200ebfcL34 and the types don't match up.. not sure why this compiles tbh

**Comment by BarneyStratford** [comment]
It should be all ready to go now.

**Pull request merged**